### PR TITLE
New combinators Cloudflow.ofCloudChannel/toCloudChannel

### PR DESCRIPTION
**Pull request by palladin**
This PR introduces two new combinators ```CloudFlow.ofCloudChannel/toCloudChannel``` and fixes the 
cancellation semantics of CloudFlow to correctly handle infinite streams of values.

### CODE REVIEW

File: src/MBrace.Flow/CloudFlow.fs [hunk 1]
```diff
@@ -1123,3 +1139,55 @@ module CloudFlow =
...
+
+                    return! combiner results
+
+                } }
+
+    /// <summary>Sends the values of CloudFlow to the SendPort of a CloudChannel</summary>
+    /// <param name="channel">the SendPort of a CloudChannel.</param>
+    /// <param name="flow">The input CloudFlow.</param>
+    /// <returns>Nothing.</returns>
+    let toCloudChannel (channel : ISendPort<'T>) (flow : CloudFlow<'T>)  : Cloud<unit> =
```
Comment (dsyme) [inline src/MBrace.Flow/CloudFlow.fs]: Consider renaming ``ISendPort`` to ``ICloudSink<_>``, and then this becomes ``toCloudSink``.

To me, a CloudChannel is an intra-cluster queue.  However, in the MBrace-suitable real-world situations I'm seeing, the vast majority of interesting sinks and sources come and go to the _outside_ world - e.g. to/from eventing systems. There are relatively few intra-MBrace queues. See also my comment here: https://github.com/mbraceproject/MBrace.Azure/issues/10#issuecomment-82265759

So I'm gradually coming to the belief that MBrace would be better and simpler if we focused on ``ICloudSink`` and ``ICloudSource`` (perhaps called ``ICloudObservable``), and avoided mentioning ``CloudChannel`` until the user really needs an intra-cluster queue, which is rare.

File: src/MBrace.Flow/CloudFlow.fs [hunk 2]
```diff
@@ -1123,3 +1139,55 @@ module CloudFlow =
...
+                        | Some n -> local { return n }
+                        | _ -> Cloud.GetWorkerCount()
+                    let! workers = Cloud.GetAvailableWorkers() 
+                    let workers = workers |> Array.sortBy (fun workerRef -> workerRef.Id)
+
+                    let createTask (collector : Local<Collector<'T, 'S>>) = 
+                        local {
+                            let! ctx = Cloud.GetExecutionContext()
+                            let! collectorf = collector
+                            let! value = CloudChannel.Receive(channel)
```
Comment (dsyme) [inline src/MBrace.Flow/CloudFlow.fs]: Is ``value`` unused?  Are unused variable warnings enabled for the project?

File: src/MBrace.Flow/CloudFlow.fs [hunk 3]
```diff
@@ -1123,3 +1139,55 @@ module CloudFlow =
...
                 } }
 
+
+
+
+                
+    /// <summary>Creates a CloudFlow from the ReceivePort of a CloudChannel</summary>
+    /// <param name="channel">the ReceivePort of a CloudChannel.</param>
+    /// <returns>The result CloudFlow.</returns>
+    let ofCloudChannel (channel : IReceivePort<'T>) : CloudFlow<'T> =
```
Comment (dsyme) [inline src/MBrace.Flow/CloudFlow.fs]: See comment on toCloudChannel.  Could this be named more general e.g. ``CloudFlow.ofCloudSource``?  

The contract for a IReceivePort (aka ICloudSource) is interesting - it starts N receive actions in parallel (where N is the degree-of-parallelism).

File: src/MBrace.Flow/CloudFlow.fs (hunk 2 above)
Comment (palladin) [inline src/MBrace.Flow/CloudFlow.fs]: Oups sorry, Thanks Don!

File: src/MBrace.Flow/CloudFlow.fs (hunk 1 above)
Comment (palladin) [inline src/MBrace.Flow/CloudFlow.fs]: ICloudSink<_> and ICloudSource<_> are certainly more generic than the ISendPort<_> and IReceivePort<_>.

File: src/MBrace.Flow/CloudFlow.fs (hunk 3 above)
Comment (palladin) [inline src/MBrace.Flow/CloudFlow.fs]: One way to specify the degree of parallelism
https://github.com/palladin/MBrace.Core/blob/master/tests/MBrace.Flow.Tests/CloudFlowTests.fs#L476
another idea is to have an overload to CloudFlow,ofCloudSource

Comment (dsyme) [inline src/MBrace.Flow/CloudFlow.fs]: Yes, for both CloudFlow.ofArray and CloudFlow.ofCloudSource it feels very implicit that there is partitioning and/or multi-readers going on.  I was going to suggest an alternative name for CLoudFlow.ofArray (like CloudFlow.paritionArray) but wasn't sure what's the best choice.  Perhaps it's ok as is.

CloudFlow.ofCloudSource is even more implicit  as people may very likely think that a cloud source is sequential, being read one-by-one. I for one didn't realize that you just planned to have many workers reading off the same channel in order to partition - that's very cool and interesting.

Comment (palladin) [inline src/MBrace.Flow/CloudFlow.fs]: Yes I agree a more explicit CloudFlow.ofCloudSource
```
CloudFlow.ofCloudSource(source : ICloudSource<'T>, degreeOfPar : int) : CloudFlow
``` 
is certainly better than the implicit.

Comment (palladin) [inline src/MBrace.Flow/CloudFlow.fs]: Maybe for this PR it is better to keep a clear seperation between the concepts of Channel(Send/Receive)Port and the new abstraction Cloud(Source/Sink). Cloud(Source/Sink) are more general in nature and a simple renaming is a bit premature.

**Pull request merged**